L1Trigger/L1TTrackMatch/plugins/L1GTTInputProducer.cc

```diff
@@ -0,0 +1,532 @@
+                 out_charge_t& chargeBit,
+                 double& expected,
+                 double& maxErrPerc,
+                 double& maxErrEpsilon,
+                 double& minErrPerc,
+                 double& minExpected) const;
+  double indexTanLambda2Eta(unsigned int indexTanLambda) const;
+  double inverseRT2InversePT(unsigned int indexRT) const;
+  void produce(edm::StreamID, edm::Event&, const edm::EventSetup&) const override;
+  template <typename T>
+  int sgn(T val) const {
+    return (T(0) < val) - (val < T(0));
+  }  // From https://stackoverflow.com/questions/1903954/is-there-a-standard-sign-function-signum-sgn-in-c-c
+  double unpackSignedValue(unsigned int bits, unsigned int nBits, double lsb) const;
+
+  // ----------member data ---------------------------
+  const edm::EDGetTokenT<TTTrackCollectionView> l1TracksToken_;
+  const std::string outputCollectionName_;
+  int debug_;
+  std::vector<out_pt_t> pt_lut_;
+  std::vector<out_eta_t> eta_lut_;
+};
+
+//
+// constructors and destructor
+//
+L1GTTInputProducer::L1GTTInputProducer(const edm::ParameterSet& iConfig)
+    : l1TracksToken_(consumes<TTTrackCollectionView>(iConfig.getParameter<edm::InputTag>("l1TracksInputTag"))),
+      outputCollectionName_(iConfig.getParameter<std::string>("outputCollectionName")),
+      debug_(iConfig.getParameter<int>("debug")) {
+  // Generate the required luts
+  generate_eta_lut();
+  generate_pt_lut();
+
+  // Define EDM output to be written to file (if required)
+  produces<TTTrackCollection>(outputCollectionName_);
+  produces<std::vector<double>>("L1GTTInputTrackPtExpected");
+  produces<std::vector<double>>("L1GTTInputTrackEtaExpected");
+}
+
+L1GTTInputProducer::~L1GTTInputProducer() {}
+
+//
+// member functions
+//
+
+/**
+  generate_lut: calculate the lut and write it to a file
+  - argument:None
+  - return (void): None
+  - method:
+    1) iterates through all possibilities of the input value
+    2) finds values of eta from input values
+    3) stores the values of eta in an array (lut)
+    4) writes out the array into a file (eta_lut.h)
+*/
+void L1GTTInputProducer::generate_eta_lut() {
+  // initialize lut array
+  eta_lut_.reserve(kEtaLutSize);
+
+  // iterate through all values in the lut
+  for (unsigned int i = 0; i < kEtaLutSize; i++) {
+    // -1 to ignore sign bit for input
+    unsigned int index = ((i + 0.5) * pow(2, (int)(kEtaInputSize - kEtaOutputSize)));
+    double newValue = indexTanLambda2Eta(index);  // map out the index to the represented eta
+    out_eta_t out_eta = newValue;                 // cast it to fxp
+    eta_lut_[i] = out_eta;                        // add value for the lut
+  }
+
+  if (debug_ >= 3) {
+    std::stringstream ss;
+    ss << "The eta_lut_[" << kEtaLutSize << "] values are ... \n";
+    for (unsigned int i = 0; i < kEtaLutSize; i++) {
+      ss << "\t" << eta_lut_[i] << "\n";
+    }
+    edm::LogInfo("L1GTTInputProducer") << "generate_eta_lut::" << ss.str();
```

Comment:
there is no need for the `std::stringstream`:
```suggestion
    edm::LogInfo log("L1GTTInputProducer");
    log << "generate_eta_lut::";
    log << "The eta_lut_[" << kEtaLutSize << "] values are ... \n";
    for (unsigned int i = 0; i < kEtaLutSize; i++) {
      log << "\t" << eta_lut_[i] << "\n";
    }
```
should be equivalent.